azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -331,78 +338,211 @@ def __init__(self, xml_text):
 
-        to_lower = lambda str_to_change: str_to_change.lower() if str_to_change is not None else None
-        ext_handler_plugin_settings = [x for x in plugin_settings if to_lower(getattrib(x, "name")) == to_lower(name)]
+        def to_lower(str_to_change): return str_to_change.lower() if str_to_change is not None else None
+        ext_handler_plugin_settings = [x for x in plugin_settings if to_lower(getattrib(x, "name")) == to_lower(handler_name)]
         if not ext_handler_plugin_settings:
             return
 
         settings = [x for x in ext_handler_plugin_settings if getattrib(x, "version") == version]
         if len(settings) != len(ext_handler_plugin_settings):
-            msg = "ExtHandler PluginSettings Version Mismatch! Expected PluginSettings version: {0} for Handler: " \
-                  "{1} but found versions: ({2})".format(version, name, ', '.join(
-                set([getattrib(x, "version") for x in ext_handler_plugin_settings]))) 
+            msg = "ExtHandler PluginSettings Version Mismatch! Expected PluginSettings version: {0} for Handler: {1} but found versions: ({2})".format(
+                version, handler_name, ', '.join(set([getattrib(x, "version") for x in ext_handler_plugin_settings])))
             add_event(AGENT_NAME, op=WALAEventOperation.PluginSettingsVersionMismatch, message=msg, log_event=False,
                       is_success=False)
-            if not settings:
-                # If there is no corresponding settings for the specific extension handler, we will not process it at all,
-                # this is an unexpected error as we always expect both versions to be in sync.
-                logger.error(msg)
-                return
-            logger.warn(msg)
-
-        runtime_settings = None
-        runtime_settings_node = find(settings[0], "RuntimeSettings")
-        seqNo = getattrib(runtime_settings_node, "seqNo")  # pylint: disable=C0103
-        runtime_settings_str = gettext(runtime_settings_node)
-        try:
-            runtime_settings = json.loads(runtime_settings_str)
-        except ValueError as e:  # pylint: disable=W0612,C0103
-            logger.error("Invalid extension settings")
-            return
+            raise ExtensionConfigError(msg)
+
+        if len(settings) > 1:
+            msg = "Multiple plugin settings found for the same handler: {0} and version: {1} (Expected: 1; Available: {2})".format(
+                handler_name, version, len(settings))
+            raise ExtensionConfigError(msg)
+
+        plugin_settings_node = settings[0]
+        runtime_settings_nodes = findall(plugin_settings_node, "RuntimeSettings")
+        extension_runtime_settings_nodes = findall(plugin_settings_node, "ExtensionRuntimeSettings")
+
+        if runtime_settings_nodes and extension_runtime_settings_nodes:
+            # There can only be a single RuntimeSettings node or multiple ExtensionRuntimeSettings nodes per Plugin
```

Comment:
I know this comes from upstream, but the names are just so confusing.

What is the difference then between RuntimeSettings and ExtensionRuntimeSettings? The latter is per extension version, not per extension, so there can be multiple per extension?

Reply:
So a quick recap on the new terminology - 

- Handler - The main Plugin/handler specified by the publishers (Eg: Microsoft.OSTC.CustomScript). 
- Extension - The runtime settings associated with a Handler (1.settings, 2.settings, etc). 
**Note:** Prior to multi-config, we used Handler and Extension interchangeably because we only supported 1 handler per extension. But moving on they would mean different things so keep an eye out for those.
- RuntimeSettings - This is the name of the node that we get from CRP which only supports **1 extension per handler** (this is the legacy behavior)
- ExtensionRuntimeSettings - This is the name of the new node that we get from CRP when the handler supports Multi-config (i.e. **multiple extensions per handler**)
- Plugin - This is the name of the node that we get from CRP that specifies the Handler. There should be 1 Plugin node per Handler

**Note:** Each plugin can have _either a single RuntimeSettings node or multiple ExtensionRuntimeSettings node, not both_.


I hope this brief overview helps! If you're still confused we can sync offline and discuss more about it :)

Reply:
Added some data to the classes we create for these. Let me know if that is enough or more data/info is required